src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1175 @@
+    my ($self, $config_rules, $config_options, $parser_options) = @_;
+
+    unless ($config_rules) {
+        *$self->{LOG}->error("$function_name: 'config_rules' argument missing (internal error)");
+        return;
+    }
+    unless ($config_options) {
+        *$self->{LOG}->error("$function_name: 'config_options' argument missing (internal error)");
+        return;
+    }
+    unless (defined($parser_options)) {
+        *$self->{LOG}->debug(2, "$function_name: 'parser_options' undefined");
+        $parser_options = {};
+    }
+
+    $self->seek_begin();
+
+    # Check that config file has an appropriate header
+    Readonly my $INTRO_PATTERN => "# This file is managed by Quattor";
+    my $intro = "# This file is managed by Quattor - DO NOT EDIT lines generated by Quattor";
+    $self->add_or_replace_lines(
+                                qr/^$INTRO_PATTERN/,
+                                qr/^$intro$/,
+                                $intro . "\n#\n",
+                                BEGINNING_OF_FILE,
+                               );
+
+    $self->_apply_rules(
+                        $config_rules,
+                        $config_options,
+                        $parser_options
+                       );
+
+    return 1;
+}
+
+
+=pod
+
+=back
+
+=head2 Private methods
+
+=over
+
+=item formatAttrValue
+
+This function formats an attribute value based on the value format specified.
+
+Arguments:
+    attr_value : attribue value
+    line_fmt : line format (see LINE_FORMAT_xxx constants)
+    value_fmt : value format (see LINE_VALUE_xxx constants)
+    value_opt : value interpretation/formatting options (bitmask, see LINE_VALUE_OPT_xxx constants)
+
+Return value:
+    A string corresponding to the value formatted according to the format specified by arguments
+    or undef in case of an internal error (missing arguments)
+
+=cut
+
+sub _formatAttributeValue
+{
+    my $function_name = "_formatAttributeValue";
+    my ($self, $attr_value, $line_fmt, $value_fmt, $value_opt) = @_;
+
+    unless (defined($attr_value)) {
+        *$self->{LOG}->error("$function_name: 'attr_value' argument missing (internal error)");
+        return undef;
+    }
+    unless (defined($line_fmt)) {
+        *$self->{LOG}->error("$function_name: 'list_fmt' argument missing (internal error)");
+        return undef;
+    }
+    unless (defined($value_fmt)) {
+        *$self->{LOG}->error("$function_name: 'value_fmt' argument missing (internal error)");
+        return undef;
+    }
+    unless (defined($value_opt)) {
+        *$self->{LOG}->error("$function_name: 'value_opt' argument missing (internal error)");
+        return undef;
+    }
+
+    *$self->{LOG}->debug(2,
+                         "$function_name: formatting attribute value >>>$attr_value<<< (line fmt=$line_fmt, value fmt=$value_fmt, value_opt=$value_opt)"
+                        );
+
+    my $formatted_value;
+    if ($value_fmt == LINE_VALUE_BOOLEAN) {
```

Comment:
a dispatch table would be better instead of  this huge if/else (or switch/case) construction. would also allow finer unittesting and easier extenstion. adding a new format, eg. by the component would mean adding new entries to the table and you could probably move some very non general code like the xrootd code out of CAF. imagine having to extend this with component specific code for each possible component. it's very non-modular this way.